### ISSUE TITLE
VCS (Version Control System) Support

### ISSUE THREAD

**Issue by kirawi** [helix-editor/helix#227]
We should have a VCS abstraction and then natively support `git` through that abstraction. Later down the line, plugins will be able to use this abstraction for other VCS. I think the abstraction should live in a new crate called `helix-vcs`.

**Comment by pickfire** [comment]
magit is interesting but it hides a lot of useful feature.

**Comment by ahkrr** [comment]
I am usingt [gitui](https://github.com/extrawurst/gitui) `About: Blazing fast terminal-ui for git written in rust`
Maybe some components could be reused, or maybe `gitui` could be called or integrated somehow?
One can already open and edit files with `VISUAL=hx EDITOR=hx gitui` from gitui.

**Comment by gbaranski** [comment]
There is also nice Neovim plugin which shows what have been changed in the source tree, https://github.com/lewis6991/gitsigns.nvim
It's similar to what vscode has.

**Comment by heliostatic** [comment]
Ah, yes, I searched `git` and missed this issue. Thanks @kirawi.

I'll add votes for:
- gutter indicators
- something with https://github.com/extrawurst/gitui/

**Comment by kirawi** [comment]
I just renamed it, so no worries :P

**Comment by QiBaobin** [comment]
One great thing is that we could reuse all our editor knowledge when providing a fugitive or magit feature, like searching, diff etc. Using another tool is a total different experience.

**Comment by poliorcetics** [comment]
Being able to use `hx` as a [`difftool`](https://git-scm.com/docs/git-difftool) in `git` (and maybe other VCS, I don't know if they have the option) would be very nice !

**Comment by the-mikedavis** [comment]
I think much of the fugitive/magit/lazygit/gitui functionality belongs outside the core as plugin(s) but I think some read-only VCS information could belong in the core. In particular diff signs (whether VCS based or not - already being implemented) but also I think it isn't a stretch to have a gutter in the core for git's blame information.

**Comment by Yevgnen** [comment]
The way how `magit` works with `transient.el` is great.

**Comment by theherk** [comment]
Just following up with a bit of [supporting data from /r/emacs](https://www.reddit.com/r/emacs/comments/wcupae/lets_share_your_top_3_packages_that_you_cant_live/). I thought it was worth noting just how high the proportion is of people who consider magit a core part of their emacs life.

It is for this reason, that I have come to believe no matter the outcome of the plugin-system-or-not conversation ongoing in #122, tight and powerful git integration should be part of the core distribution of helix. Maybe @extrawurst, the creator of [gitui](https://github.com/extrawurst/gitui) would be interested in joining such a cause.

**Comment by babakabdolahi** [comment]
> Just following up with a bit of [supporting data from /r/emacs](https://www.reddit.com/r/emacs/comments/wcupae/lets_share_your_top_3_packages_that_you_cant_live/). I thought it was worth noting just how high the proportion is of people who consider magit a core part of their emacs life.
> 
> It is for this reason, that I have come to believe no matter the outcome of the plugin-system-or-not conversation ongoing in #122, tight and powerful git integration should be part of the core distribution of helix. Maybe @extrawurst, the creator of [gitui](https://github.com/extrawurst/gitui) would be interested in joining such a cause.

I also think git integration is also as fundamental as LSP, grep and fuzzy finder and should be part of the core distribution of helix.

**Comment by wojpawlik** [comment]
The simplest and most unixy way to go would be to add a keybind that saves all files and launches https://github.com/extrawurst/gitui

**Comment by theherk** [comment]
Maybe that would be the most simple to implement, but I'm not convinced it would be the most simple to use, which is also important. I think there is some benefit to unifying the experience, so the interface and behavior are consistent.

In addition, the time machine and blame interfaces within the editor is really a nice quality of life tool, that just wouldn't work if the git tool were completely isolated from the editor.

**Comment by theherk** [comment]
Also, I saw that @extrawurst just starred helix, so maybe they'll want to get involved or offer some suggestions on how tightly they could / should be coupled. 🤞

**Comment by univerz** [comment]
\+ diffs with editor-quality syntax highlighting and the ability to do final cleanups / small fixes right in the side-by-side diff are solid productivity boosts.

**Comment by dahmc** [comment]
I really want to try helix but not having first class git integration is the one and only blocker for me. 

Since it already implements LSP it should be clear it's aimed at source code editing, and therefore I'd rather have core support than plug-in support.

**Comment by vkoskiv** [comment]
My Rust is sadly far too weak to take a stab at this myself, but the ability to hit something like `<space>-b` to show a git blame for the current line, and/or multiple lines in a selection would be just excellent. I think it would be a great fit for Helix, and perhaps a somewhat modest starting point for broader Git integration in the future.

**Comment by gobijan** [comment]
This is the best keyboard based Git integration for an editor (Sublime) I came across. 
Could be a nice inspiration.

https://github.com/timbrel/GitSavvy

**Comment by quantonganh** [comment]
> but the ability to hit something like <space>-b to show a git blame for the current line, and/or multiple lines in a selection would be just excellent.

@vkoskiv If you are using [WezTerm](https://wezfurlong.org/wezterm/index.html), the current file name and line number can be parsed from the status line by using something like this:

```sh
status_line=$(wezterm cli get-text | rg -e "(?:NORMAL|INSERT|SELECT)\s+[\x{2800}-\x{28FF}]*\s+(\S*)\s[^│]* (\d+):*.*" -o --replace '$1 $2')
filename=$(echo $status_line | awk '{ print $1}')
line_number=$(echo $status_line | awk '{ print $2}')
```

then it can be passed to [tig](https://jonas.github.io/tig/):

```sh
case "$1" in
  "blame")
    split_pane_down
    echo "tig blame $filename +$line_number" | $send_to_wezterm_pane
    ;;
```

```sh
split_pane_down() {
  pane_id=$(wezterm cli get-pane-direction down)
  if [ -z "${pane_id}" ]; then
    pane_id=$(wezterm cli split-pane)
  fi

  wezterm cli activate-pane-direction --pane-id $pane_id down

  send_to_wezterm_pane="wezterm cli send-text --pane-id $pane_id --no-paste"
}
``` 

and finally, you can create a new minor mode bound to `;`:

```toml
[keys.normal.";"]
b = ":sh hx-wezterm.sh blame"
```

**Comment by dlants** [comment]
I agree that version control is something that's really fundamental, and based on the approach of helix so far seems like it belongs in the core.

I also think it would be better to integrate it into helix itself in situations where you want to select, yank or edit text. (rather than delegating to something like gitui). So things like browsing diffs, blame, looking at previous versions of files, etc...

The rest of the stuff (staging, commits, remotes, git log) is maybe less important to do inside the editor itself, and could be delegated to a plugin or a separate tool?

I really enjoy using fugitive, but I do think that it's not quite helix-y from the point of view of discoverability. I've been using it for like 20 years and I still only use a really small subset of what I know it's capable of, since I forget how to activate the fancier features and never bother to look them up, resorting to the git cli or web UI instead.

**Comment by ds-281** [comment]
The one killer feature I'm really missing from `vim-fugitive` is the reblame function of `:Git blame` (navigating on a specific line and pressing `~` reloads the file as in that commit's tree with its own blame) and the possibility to load the selected commit in a dedicated pane (with `o`).

This is really the only missing thing in my workflow to leave `vim` behind.

**Comment by dmitry-krasilnikov** [comment]
I am willing to work on `git blame` feature starting with a simple "blame this line and show it in a popup" and then improving on that, e.g. full blame of a file, re-blame. I already have "git blame line" implemented as a command in my fork of a helix and it works well. It's far from being ready as a PR, the architecture is not right, and there are no tests, but I can re-write it and open a PR if it's welcome.

**Comment by jpaju** [comment]
If I've understood correctly, the problem is that Helix uses [gix](https://github.com/Byron/gitoxide) for git integration and gix (to my knowledge) doesn't have `git blame` implemented yet

**Comment by extrawurst** [comment]
@jpaju you are right it currently does not support blame but there is some convo around implementing it, let’s make sure to not duplicate efforts here. You can find more context here: https://github.com/Byron/gitoxide/issues/1078#issuecomment-2211771332

**Comment by webdev23** [comment]
This does work.

[lazygit](https://github.com/jesseduffield/lazygit) and create a config file:

`hx ~/.config/lazygit/config.yml`

```
os:
  edit: "echo {{filename}} > /tmp/lazygit_out.txt && killall lazygit"
  editInTerminal: true
```

Helix config.toml:

```
[keys.normal]
"C-g" = [
    ':set mouse false',
    ':insert-output lazygit',
    ':redraw',
    ':set mouse true',
    ':open /tmp/lazygit_out.txt',
    'select_all',
    'split_selection_on_newline',
    'goto_file',
    ':buffer-close! /tmp/lazygit_out.txt',
    ':sh rm -f /tmp/lazygit_out.txt',
    ':rla'
]
```

Usage example:
Restore a file from an older commit:

    Within helix, Ctrl + g opens lazygit, choose a branch on the left, or press Ctrl + s to search by filename.

    Select the file in the tree, and press c (Checkout)

    Press e to open the editor (helix in background).

File is restored in the workspace, helix is open on the file or the buffer is refreshed if the file was already open/ahead, various tweaks are possible from there.
